### DISCUSSION TITLE
Problem solving in Factor

### DISCUSSION THREAD

**Discussion by Bubbler-4** [factor/factor#2391] [General]
As I already mentioned, I'm uploading solutions to various problems to [Bubbler-4/factor-problem-solving](https://github.com/Bubbler-4/factor-problem-solving). I have some questions and a few feature requests:

### Questions

- If I want to contribute these solutions to Factor repository, are there any source code guidelines to follow?
- I'm uploading a test file and a docs file for every single solution. Are there guidelines specifically for tests and docs?
- If I want to explain the approach in the help, should I write it in individual word help or in the article section?
- I'm using the problem code (e.g. `fib` for the [problem `FIB`](http://rosalind.info/problems/fib/)) as the word name for Rosalind solutions. One issue with this is that some problem codes clash with built-in words, such as `gc`. Is it OK to keep the names as-is?

### Feature requests

- Programming tasks often work with individual characters, and writing out `CHAR: x` for every single char that matters is pretty cumbersome to my experience. One example is [aoc2020.day12](https://github.com/Bubbler-4/factor-problem-solving/blob/6cbc118a0e3e3fa46f71a26a3876dc8724b743c0/aoc2020/day12/day12.factor#L8), where one has to select one of seven quotations by the given char.
- A minor issue when working with 2D arrays ([aoc2020.day11](https://github.com/Bubbler-4/factor-problem-solving/blob/main/aoc2020/day11/day11.factor)) was the lack of library word that takes row and column coordinates and extracts one item from a matrix. Given that Factor does not support anything close to `array[row][col]` syntax, having something like `deep-nth` or `matrix-mnth` would be nice.

**Comment by jonenst** [comment]
> If I want to contribute these solutions to Factor repository, are there any source code guidelines to follow?

We have solutions to project-euler in the repo, so I think aoc solutions and rosalind solutions should be fine too ? Though we have to stop at some point..

> I'm uploading a test file and a docs file for every single solution. Are there guidelines specifically for tests and docs?

Tests are very important :) they are present for project euler solutions for example: extra/project-euler/001-tests.factor 
For vocabs like that, docs are not mandatory I would say.

>    If I want to explain the approach in the help, should I write it in individual word help or in the article section?

You can do what you want for explanations, my preferred approach would be directly code comments but everything works I guess.

> I'm using the problem code (e.g. fib for the problem FIB) as the word name for Rosalind solutions. One issue with this is that some 
problem codes clash with built-in words, such as gc. Is it OK to keep the names as-is?
It's ok to clash with core words in such vocabs I guess if you want to. Clashing with core words is not special, it's just that they are used more often so it's more annoying, but it's handled just like every other clash.


> Programming tasks often work with individual characters, and writing out CHAR: x for every single char that matters is pretty cumbersome to my experience. One example is aoc2020.day12, where one has to select one of seven quotations by the given char.

I think it's ok.. I feel like something like this was already suggested, maybe with erg's new parser ? Don't remember.

> A minor issue when working with 2D arrays (aoc2020.day11) was the lack of library word that takes row and column coordinates and extracts one item from a matrix. Given that Factor does not support anything close to array[row][col] syntax, having something like deep-nth or matrix-mnth would be nice.

I agree, I must have implemented these at least 50 times now :p There's something coming up in array.shaped though, haven't used it yet.

**Reply by Bubbler-4** [reply]
Hmm. I guess I'll keep my solutions in my repo for now.

`array.shaped` is definitely interesting. I haven't tried it out either, but it conceptually looks like an APL array, which should come very handy for many purposes if fully implemented.

**Comment by jonenst** [comment]
Also, side note, I'm looking at your solutions every day, congrats they are really nice !

**Comment by mrjbq7** [comment]
We’ve talked about character syntax a fair amount. I’m not sure yet what’s the best option, some ideas include:

- single quote like Java 'ch'
- backtick `ch
- shorter tag C: ch
- lowercase tag char: ch
- strings of length 1 like Python

Then there is a whole conversation about code point characters versus graphemes based characters. 

So far it hasn’t been so restrictive, albeit I would suggest this is a nice area to improve as you pointed out. If we use single quotes, that has the potential downside of removing that syntax from other possible use cases including using it the way Python does. 

Love to continue your lessons learned through the rest of the month. 

Regarding problem solving in the core, I like having one giant repo. At some point maybe we have to split it out into a separate official project repo under the Factor organization.  However, it’s great to run the tests and use the solution examples to improve our standard library.

**Reply by Bubbler-4** [reply]
One idea just came to my head: how about `C"ch"`, using `C"` as a "string" opener?

I'll try to continue problem solving as much as I can. On my pending list are past AoC events and Project Euler (I have a PE account with 200+ solves btw, and I'll be happy to re-solve them in Factor).

**Comment by Bubbler-4** [comment]
[All Factor solutions for AoC 2020](https://github.com/Bubbler-4/factor-problem-solving/tree/main/aoc2020)!

The original problems can be found [here](https://adventofcode.com/2020) (though you need to log in/create an account and solve part 1 of each problem to access part 2). My solutions also have help files containing the brief summary of each problem, and the approach used to solve it (for mostly harder ones; spoilers ahead).
